botocore/utils.py

```diff
@@ -245,25 +240,106 @@ def set_value_from_jmespath(source, expression, value, is_first=True):
 
 class InstanceMetadataFetcher(object):
 
-    _REQUIRED_CREDENTIAL_FIELDS = [
-        'AccessKeyId', 'SecretAccessKey', 'Token', 'Expiration'
-    ]
+    class _RetriesExceededError(Exception):
+        """Internal exception used when the number of retries are exceeded."""
+        pass
 
     def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
-                 num_attempts=1, url=METADATA_SECURITY_CREDENTIALS_URL,
-                 env=None, user_agent=None):
+                 num_attempts=1, base_url=METADATA_BASE_URL,
+                  env=None, user_agent=None):
         self._timeout = timeout
         self._num_attempts = num_attempts
-        self._url = url
+        self._base_url = base_url
         if env is None:
             env = os.environ.copy()
         self._disabled = env.get('AWS_EC2_METADATA_DISABLED', 'false').lower()
         self._disabled = self._disabled == 'true'
         self._user_agent = user_agent
         self._session = botocore.httpsession.URLLib3Session(
             timeout=self._timeout,
-            proxies=get_environ_proxies(self._url),
+            proxies=get_environ_proxies(self._base_url),
+        )
+
+    def get_request(self, url_path, custom_retry=None):
```

Comment:
I think for the signature here we just expose a `needs_retry` that overrides whatever the general `_needs_retry` method. It is a little weird that no matter what a user of this interface cannot override the default empty/non-200 retry logic.